File: Point3D.java
public class Point3D
{
	private int x, y, z;
	public Point3D(int x, int y, int z)
	{
		this.x = x;
		this.y = y;
		this.z = z;
	}
	
	public int getX()
	{
		return x;
	}
	
	public int getY()
	{
		return y;
	}
	
	public int getZ()
	{
		return z;
	}
	
	@Override
	public String toString()
	{
		return String.format("(%d, %d, %d)", x, y, z);
	}
	
	public double getEuclidianDistance(Point3D other)
	{
		double xdiff = x - other.x;
		double ydiff = y - other.y;
		double zdiff = z - other.z;
		return Math.sqrt(xdiff * xdiff + ydiff * ydiff + zdiff * zdiff);
	}
	
	public double getManhattanDistance(Point3D other)
	{
		double xdiff = x - other.x;
		double ydiff = y - other.y;
		double zdiff = z - other.z;
		return Math.abs(xdiff) + Math.abs(ydiff) + Math.abs(zdiff);
	}
	
	public double getChebyshevDistance(Point3D other)
	{
		double xdiff = Math.abs(x - other.x);
		double ydiff = Math.abs(y - other.y);
		double zdiff = Math.abs(z - other.z);

		// return max(xdiff, ydiff, zdiff).
		if (xdiff > ydiff)
		{
			if (xdiff > zdiff)
				return xdiff;
			return zdiff;
		}
		if (ydiff > zdiff)
			return ydiff;
		return zdiff;
	}
	
	@Override
	public boolean equals(Object o)
	{
		if (o instanceof Point3D)
		{
			Point3D other = (Point3D)o;
			return x == other.x && y == other.y && z == other.z;
		}
		if (o instanceof PointF3D)
		{
			PointF3D other = (PointF3D)o;
			return x == other.getX() && y == other.getY() && z == other.getZ();
		}
		return false;
	}
	
	@Override
	public int hashCode()
	{
		return (x << 20) | (y << 10) | z;
	}
}

File: src/Point3D.cpp
//
// Created by Geoff M. on 11/8/17.
//

#include "include/Point3D.h"
#include <cmath>

int Point3D::getX() const { return x; }
int Point3D::getY() const { return y; }
int Point3D::getZ() const { return z; }

Point3D::Point3D(int x, int y, int z)
{
    this->x = x;
    this->y = y;
    this->z = z;
    isEmpty = false;
}

Point3D::Point3D()
{
    isEmpty = true;
}

Point3D Point3D::operator+(const Point3D other) const
{
    //__glibcxx_assert(!isEmpty && !other.isEmpty);
    return Point3D(x + other.x,
                   y + other.y,
                   z + other.z);
}

Point3D Point3D::operator-(const Point3D other) const
{
    //__glibcxx_assert(!isEmpty && !other.isEmpty);
    return Point3D(x - other.x,
                   y - other.y,
                   z - other.z);
}

int Point3D::manhattanNorm()
{
    return (int)(std::abs(x) + std::abs(y) + std::abs(z));
}

bool Point3D::operator ==(const Point3D other) const
{
    if (isEmpty ^ other.isEmpty)
        return false;
    return (x == other.x) && (y == other.y) && (z == other.z);
}

bool Point3D::operator ==(const state other) const
{
    if (isEmpty)
        return false;
    return (x == other.x) && (y == other.y);
}

bool Point3D::operator !=(const Point3D other) const
{
    return (x != other.x) || (y != other.y) || (z != other.z) || (isEmpty != other.isEmpty);
}

bool Point3D::isUninitialized() const
{
    return isEmpty;
}